### DISCUSSION TITLE
Component guide - How to create components with reactive attributes in an idiomatic vanjs way

### DISCUSSION THREAD

**Discussion by iongion** [vanjs-org/van#23] [Q&A] [answered]
Some help in this regard might be needed.

There is this example that shows a `style` that is reactive

```javascript
const dom3 = div(
  {style: {deps: [counter], f: c => `font-size: ${c}em;`}},
  "Text")
```

- What if one wants to create a custom attribute that is not `style` but for example `data-icon` in a fictive `Icon` component ?
- What if one wants to create a reactive `isOpen` boolean attribute for a fictive popup / modal / collapsible ?

I am still wrapping my head with the reactivity model to try to get-it, but it is so fun!

An example

```typescript
export interface ButtonProps<E extends HTMLButtonElement | HTMLAnchorElement | HTMLElement = HTMLButtonElement> extends ActionProps<E> {
  children?: ChildDom[];
  icon?: IconName | Node | [];
  rightIcon?: IconName | Node | [];
  alignText?: Alignment;
  intent?: Intent;
  active?: boolean | State<boolean>;
  fill?: boolean | State<boolean>;
  large?: boolean | State<boolean>;
  loading?: boolean | State<boolean>;
  minimal?: boolean | State<boolean>;
  outlined?: boolean | State<boolean>;
  small?: boolean | State<boolean>;
  type?: "submit" | "reset" | "button";
}

const { button, span } = van.tags;

export const Button = (props: ButtonProps = {}, ...rest: readonly ChildDom[]) => {
  const componentClasses = classNames(
    // default class
    Classes.BUTTON,
    // reactive
    props.active ? Classes.ACTIVE : null,
    props.minimal ? Classes.MINIMAL : null,
    props.outlined ? Classes.OUTLINED : null,
    props.fill ? Classes.FILL : null,
    props.small ? Classes.SMALL : null,
    props.alignText ? AlignmentClassMap[props.alignText] : null,
    props.intent ? IntentClassMap[props.intent] : null
  );
  const createText = () => {
    return props.text ? span({ class: Classes.BUTTON_TEXT }, props.text) : [];
  };
  const createIcon = (icon?: IconName | Node | []) => {
    return icon ? Icon({ icon }) : [];
  };
  return button(
    // attributes
    {
      class: componentClasses,
      onclick: (e: MouseEvent) => {
        if (props?.onClick) {
          props.onClick(e);
        }
      },
    },
    // children
    createIcon(props?.icon),
    ...(props?.children || []).concat(rest),
    createText(),
    createIcon(props.rightIcon)
  );
};
Button.displayName = `${DISPLAYNAME_PREFIX}.Button`;
```

How can one make reactive `componentClasses` ?

I've tried this hereunder but because props are either `scalar` or `State<scalar>` I can't get it to work unless the user makes all `State<scalar>`

```typescript
  const componentClasses = {
    deps: [props.active, props.minimal, props.outlined, props.fill, props.small, props.alignText, props.intent],
    //
    (active, minimal, outlined, fill, small, alignText, intent) => classNames(
      // default class
      Classes.BUTTON,
      // reactive
      active ? Classes.ACTIVE : null,
      minimal ? Classes.MINIMAL : null,
      outlined ? Classes.OUTLINED : null,
      fill ? Classes.FILL : null,
      small ? Classes.SMALL : null,
      alignText ? AlignmentClassMap[props.alignText] : null,
      intent ? IntentClassMap[props.intent] : null
  )
};
```

**Comment by Tao-VanJS** [comment, marked as answer]
> What if one wants to create a custom attribute that is not style but for example data-icon in a fictive Icon component ?

Custom attributes of native DOM elements are supported in the same way as properties. When the field of `props` arg doesn't exist as an element property, VanJS will set the `prop` value as a DOM attribute. This behavior is discussed at https://vanjs.org/advanced#dom-attributes-vs-properties

> What if one wants to create a reactive isOpen boolean attribute for a fictive popup / modal / collapsible ?

For custom components (which is just a plain JavaScript function), there is no builtin support for state-derived properties, although you can provide similar support in the implementation of the component function. I have a feeling that for most use cases of custom components, state-derived properties might not be that necessary. Unlike native DOM elements, where some properties need to be provided in a specific way as defined in HTML specification (for instance, `style` property has to be in the format of `font-size: ${c}em;`), custom components can define the `props` in a way to avoid the need of state-derived properties.

That said, if you feel that you have a strong use case where state-derived properties are necessary for custom components, you can let me know. I will be curious to know if there is a straightforward solution to your problem. Meanwhile, the real-world use cases provide great feedback for the evolution of VanJS

**Reply by iongion** [reply]
Thank you so much for the detailed response, I do am having a lot of fun, it is finally, just JS and good old data-binding, "no magic"!

I have this opinion, as long as you expose custom properties, it is somehow intuitive to assume they are reactive, they are the public interface of a component. But this does not mean that they always have to be reactive, that is why I was trying this `active?: boolean | State<boolean>;` kind of definition.

I do like the interface of the default tags

```typescript
(first?: Props | ChildDom, ...rest: readonly ChildDom[])
```

I tried to follow it except the first argument, I want to achieve that too

```typescript
export const Button = (props: ButtonProps = {}, ...rest: readonly ChildDom[])
```

this part is especially interesting for me, I am trying to understand it, see if I can have something similar for the components architecture I am experimenting with

```typescript
export type State<T> = {
  val: T
  onnew(l: (val: T, oldVal: T) => void): void
}

// Defining readonly view of State<T> for covariance.
// Basically we want State<string> implements StateView<string | number>
export interface StateView<T> {
  readonly val: T
}

export type Primitive = string | number | boolean | bigint

export type PropValue = Primitive | Function | null

export interface DerivedProp {
  readonly deps: readonly StateView<unknown>[]
  readonly f: (...args: readonly any[]) => PropValue
}

export interface Props {
  readonly [key: string]: PropValue | StateView<PropValue> | DerivedProp
}

export type ChildDom = Primitive | Node | StateView<Primitive | null | undefined> | readonly ChildDom[]

export type TagFunc<Result> = (first?: Props | ChildDom, ...rest: readonly ChildDom[]) => Result
```

**Reply by Tao-VanJS** [reply]
> I have this opinion, as long as you expose custom properties, it is somehow intuitive to assume they are reactive, they are the public interface of a component. But this does not mean that they always have to be reactive, that is why I was trying this active?: boolean | State<boolean>; kind of definition.

👍 I agree with you. I believe this should be a good interface if you want to export the custom component to a wide range of users. One caveat is that for custom properties with type `T | State<T>` that are not used as child DOMs or prop values, the shapes of `T` and `State<T>` are indeed different. Component authors need to be aware of that.

**Reply by iongion** [reply]
If you can give some guidance

```typescript
interface IconProps {
  icon: string | StateView<string>;
}
const Icon = ({ icon }: IconProps, ...rest) => {
  // How do I do logic when icon is not a string but a state var and it changes ? I still want to support when icon is string only
  return div({ class: "Icon", "data-icon": icon });
};

interface ButtonProps extends IconProps {}
const Button = ({ icon }: ButtonProps, ...rest) => {
  return button({ class: "Button" }, Icon({ icon });
};

// Use case 1 - icon is a string primitve
const App = () => {
   return div(
     {class: "App"},
     Button({ class: "MyButton", icon: "my-icon" })
   );
}

// Use case 2 - icon is a StateView<string>
const App = () => {
   const icon = van.state("my-icon");
   return div(
     {class: "App"},
     Button({ class: "MyButton", icon: icon, onclick: () => icon.val = "other-icon" })
   );
}
```

What I have done so far is something like this

Introduced a helper

```typescript
export const isState = (val: unknown) => Object.getPrototypeOf(val) == Object.getPrototypeOf((van as any)["state"]());
```

Then refactored the Icon to be like this

```typescript

export enum IconName {
  REFRESH = "refresh",
  DUPLICATE = "duplicate",
  DATABASE = "database",
  FUNCTION = "function",
  COG = "cog",
  INFO_SIGN = "info-sign",
  CARET_DOWN = "caret-down",
  CARET_UP = "caret-up",
  CARET_RIGHT = "caret-right",
}

export type PIcon = IconName | StateView<IconName> | Node | [];

export interface IconProps extends Props {
  icon: PIcon;
  children?: ChildDom[];
}

export function Icon(props: IconProps, ...rest: readonly ChildDom[]) {
  const icon = props.icon;
  const children = [...(props?.children || []).concat(rest)];
  if (icon instanceof Node) {
    return icon;
  }
  // case 2 - icon is a van.state
  if (isState(icon)) {
    return span(
      // attrs
      {
        "data-icon": icon as StateView<IconName>,
        "aria-hidden": true,
        class: {
          deps: [icon as any],
          f: (iconName) => classNames(Classes.ICON, `${Classes.ICON}-${iconName}`, props.class),
        },
      },
      children
    );
  }
  // case 1 - icon is a string primitive (from the enum)
  let iconName = icon.toString();
  const iconClass = `${Classes.ICON}-${iconName}`;
  const iconClasses = classNames(Classes.ICON, iconClass, props.class);
  const attrs = {
    "data-icon": iconName,
    "aria-hidden": true,
    class: iconClasses,
  };
  return span(attrs, children);
}
```

1. What I like is that Icon can be a string, a state var and even a node (which in turn returns itself)
2. What I don't like is the duplication of code between case 1 and case 2

Is there a better way to achieve this ? What if `class` attribute of the icon can also be a `StateView<string>`, how many ifs should I write ? What If more `optional` primitive or `StateView<primitive>` are exposed, this method above wouldn't scale that well.

Thank you!

**Reply by Tao-VanJS** [reply]
I feel that the most straightforward way is to have a `toState()` function which converts non-state objects to states but is noop to states. It might make sense to add this function to core `van.js` in a future release but before that, you can implement one as a workaround. WDYT?

**Reply by iongion** [reply]
That could work, wouldn't boxing them incur performance cost ?
Something like this

```typescript
import van from "@/van-0.11.10.debug";
import type { Primitive } from "@/van-0.11.10.debug";

export const isState = (val: unknown) => Object.getPrototypeOf(val) == Object.getPrototypeOf((van as any)["state"]());
export const toState = (val: Primitive) => (isState(val) ? val : van.state(val));
```

actually more like this

```typescript
import van from "@/van-0.11.10.debug";
import { State, Primitive, PropValue, StateView, DerivedProp } from "@/van-0.11.10.debug";

export function isState(val: Primitive | StateView<Primitive | null | undefined> | DerivedProp | null | undefined) {
  if (val === undefined || val === null) return false;
  return Object.getPrototypeOf(val) == Object.getPrototypeOf((van as any)["state"]());
}
export function toState(val: Primitive | StateView<Primitive | null | undefined> | DerivedProp | null | undefined) {
  return isState(val) ? (val as State<PropValue>) : van.state(val);
}
```

you can see it in action at <https://stackblitz.com/edit/vitejs-vite-kc1gr9>

**Reply by Tao-VanJS** [reply]
Actually, another idea to avoid the boxing overhead:

1. `getVal` to get the `val` property for states, no-op for others.
2. `van.bind` supports both states and plain values, where plain values are equivalent to states that never change.

**Reply by iongion** [reply]
You mean this `val` from the state ?

```typescript
export type State<T> = {
  val: T
  onnew(l: (val: T, oldVal: T) => void): void
}
```

How would that work, a sample implementation that would work for both State<T> objects and plain/scalar values ?

Huge thank you!


Actually boxing is not that ugly - look at this example of a collapsible component

1. Created `PType` wrappers for common primitives - this works well with enums too
2. Ability to express default typesafe prop values (I actually tried to use deepmerge npm but it flattens out Proxy objects so I can only support shallow merge / replace) - Having some trouble with props that are nested objects, basically not supported, but which could be expected (for example one would want to define only the dayPickerProps in a DatePicker component)
3. Expressing deps and the reactor function is explicit, which is nice, no more 1K confusing `useEffect`s 🤣 
4. Expressing deps is verbose - someone brighter might use toys such as `XState` to describe the real state machine that a component is, relying on vanjs plus a human friendly syntax.

I do like how it shaped up - yeah yeah, reinventing the wheel, so what ? 😆 

```typescript
type Prop<T> = T | StateView<T>;
type PBoolean = Prop<boolean>;

interface ICollapseProps extends Props {
  isOpen?: PBoolean;
  keepChildrenMounted?: PBoolean;
}
const defaultProps: ICollapseProps = {
  isOpen: false,
  keepChildrenMounted: false,
};
export type CollapseProps = ICollapseProps;
export function Collapse(userProps: CollapseProps = defaultProps, ...rest: readonly ChildDom[]) {
  const props = {
    ...defaultProps,
    ...userProps,
  };
  return div(
    {
      class: {
        deps: [toState(props.isOpen), toState(props.class)],
        // reactor
        f: (isOpen, className) => classNames(Classes.COLLAPSE, { [Classes.COLLAPSE_OPEN]: isOpen }, className),
      },
    },
    div({ class: Classes.COLLAPSE_BODY }, rest)
  );
}
Collapse.displayName = `${DISPLAYNAME_PREFIX}.Collapse`;
```

**Reply by Tao-VanJS** [reply]
I'm going to create an issue to describe the proposal that reflects my current thinking in details. I think the basic idea is pretty similar to what you described in your sample code.

With the official support, you don't need to write:

```js
{
  deps: [toState(props.isOpen), toState(props.class)],
  ...
}
```

Instead, you can simply have:

```js
{
  deps: [props.isOpen, props.class],
  ...
}
```

Because in the future version of VanJS, state and non-state objects will be handled uniformly in `van.bind`.

Before the feature is implemented, adding a `toState` call is a reasonable workaround, which can be removed in a one-time cleanup once the feature is released.

**Reply by iongion** [reply]
Sounds great. In my experiment with blueprint, props can also be Node, hope that will handle this case too and keep reactivity.

I have another learning, while implementing / porting the ProgressBar

See here <https://stackblitz.com/edit/vitejs-vite-kc1gr9?file=src%2Fblueprint%2Fcomponents%2FProgressBar.ts>

Due to bindings mechanism, I have to duplicate the clamping / percentage, first is for the `aria-valuenow` attribute and second is for the `width` of the nested `<div>` with class `PROGRESS_METER` - this is really computing twice, which will cost with scale.

In the original typescript/react component, they do it only once in the `render` function

<https://github.com/palantir/blueprint/blob/ffc88eef61d1662bdba1286a2e0fe7d7eb0041c4/packages/core/src/components/progress-bar/progressBar.tsx#L66>

Looking for some ideas to avoid that duplication

**Reply by Tao-VanJS** [reply]
#38 is created to track the implementation of supporting both state and non-state objects in `van.bind` and state-derived properties.

Regarding the duplication between clamping and percentage, the key missing feature is "derived states". In earlier versions of VanJS, derived states were supported. However, the support was removed due to the issues of [GC](https://vanjs.org/advanced#gc) in very complex bindings. Indeed, implementing a correct GC behavior is extremely difficult (if not impossible) to get it right without significantly increasing the complexity and the size of the framework. So I gave up this feature and introduced [`onnew` event handler](https://vanjs.org/tutorial#public-interface-of-state-objects) in state objects to handle the inter-dependencies between state objects. With `onnew` we're able to propagate changes between states but users need to be careful about GC, see: https://vanjs.org/advanced#onnew-listeners-are-not-subject-to-gc

Thus for the situation, generally we have 3 options:
1. Accept the duplicate calculation, which is probably okay if the calculation is not part of the performance bottleneck.
2. Use `van.bind` to bind the entire DOM node, where in the generation function, we can calculate the 2 properties altogether, thus eliminating the duplicate computation. But the down side of this approach is whenever the state gets updated, the DOM node needs to be replaced by a newly generated one, which can impact the performance as well.
3. Use `van.bind` with [stateful binding](https://vanjs.org/tutorial#stateful-binding), this way we can implement a customized binding logic to avoid DOM node being regenerated when property values need to be updated. This approach has the most optimal performance characteristics but needs some special handling in the implementation.

Admittedly, from users' perspective, React offers a more convenient user interface for this specific situation. I totally admitted that for complex UI applications VanJS is not on par with React. I don't position VanJS as React's replacement or competitor. I am rational enough to recognize the fact. VanJS is just a project that is built with only my personal time. There is no point it can compete with React which has hundred-billion-dollar companies like Meta's backing. The purpose of VanJS, is to solve the problem where other mainstream frameworks don't seem quite care to solve, which is:

_I know very little about front end programming. I have no comprehensive tools, nor I am familiar with them. How can I build a useful UI app within just a couple hours?_

**Reply by Tao-VanJS** [reply]
@iongion, I just realized a perhaps more effective way to avoid the duplication - using `onnew` event handler to register derived states. In your example, you can define a derived state `percent`, whose `val` property is changed whenever the source state `value` is changed, with the code like below:

```js
const percent = van.state(0)
value.onnew(v => percent.val = value == null ? undefined : 100 * clamp(value, 0, 1))
```

One caveat is to avoid generating `ProgressBar` component many times with the same state object `value`, because the `onnew` event handler never gets GC-ed (see [this](https://vanjs.org/advanced#onnew-listeners-are-not-subject-to-gc)). You can consider to use a new state object for `value` whenever `ProgressBar` is regenerated.

**Reply by Tao-VanJS** [reply]
Hi @iongion,

With the release of VanJS 1.0, we have much better support for derived states.

For the issue being discussed, you can simply write now:

```js
const value = van.state(...)
...

const percent = van.derive(() => value.val == null ? undefined : 100 * clamp(value.val, 0, 1))
```

[`van.derive`](https://vanjs.org/tutorial#derived-state) is a powerful API that can enable lots of programming idioms.

Hope it helps :-)

**Reply by efpage** [reply]
> **@iongion:** _I just realized a perhaps more effective way to avoid the duplication - using onnew event handler to register derived states._ 

States are not changing without reason. There is always an event, that causes a state change. And in VanJS, any state change is directly propagated to all ressources that are bound to a state, which includes the onnew-event. So, onNew works pretty much like a direct event handler.

Using Van-States is just an indirection that may be helpful to organize your code. But in your case it is possibly easier to call the event handler directly. If you want to avoid direct calls, you can add an onNew property to the calling routine like this:

```JS
let callerOnNew = null
function caller{
   ... someCode...
  if (callerOnNew) 
     callerOnNew()
}

function myHandler() {
   ...yourHandlerCode...
}
callerOnNew = myHandler  // --> activate the handler
```
